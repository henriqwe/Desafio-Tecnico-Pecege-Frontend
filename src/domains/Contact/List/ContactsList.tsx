import type { IContact } from '&types/contact'
import { fetchContacts } from '&services/queries/fetchContacts'
import { useEffect, useState } from 'react'
import { sortArrayByKey } from '&utils/handles/sortArrayByKey'
import { useNavigate } from 'react-router-dom'
import { routes } from '&utils/routes'
import { ContactCard } from '&components/ContactCard/ContactCard'
import * as S from '&domains/Contact/List/style'
import { IllustrationCard } from '&domains/Contact/List/IllustrationCard'
import { useQuery, useQueryClient } from '@tanstack/react-query'
import { FloatingActionButton } from '&components/FloatingActionButton/FloatingActionButton'

interface props {
  ascendentName: boolean
  nameValue?: string
}

export function ContactList({ ascendentName, nameValue }: props) {
  const navigate = useNavigate()
  const [contactsFiltered, setContactsFiltered] = useState<IContact[]>()
  const queryClient = useQueryClient()

  async function handleFetch() {
    const contacts = queryClient.getQueryData(['contacts']) as IContact[]

    if (contacts) {
      return contacts
    }
    return fetchContacts()
  }
  const contactsQuery = useQuery({
    queryKey: ['contacts'],
    queryFn: () => handleFetch(),
    refetchOnWindowFocus: false,
    retry: false,
    suspense: true
  })

  const showIllustration =
    (contactsFiltered?.length === 0 ||
      contactsQuery?.data?.length === 0 ||
      contactsQuery?.data === undefined) &&
    !contactsQuery?.isLoading

  function onFilter(text?: string) {
    if (!text) {
      return contactsQuery?.data as IContact[]
    }
    const newDataList = contactsQuery?.data?.filter((contact) =>
      contact.name.toLocaleLowerCase()?.includes(text.toLocaleLowerCase())
    )
    return newDataList as IContact[]
  }

  function handleData(data: IContact[], asc: boolean) {
    setContactsFiltered(sortArrayByKey({ data, asc, key: 'name' }))
  }

  function handleNavigate(id: number) {
    navigate(routes.viewContact.path(id.toString()))
  }

  useEffect(() => {
    if (contactsQuery?.data) {
      const data = onFilter(nameValue)
      handleData(data, ascendentName)
    }
  }, [contactsQuery?.data, ascendentName, nameValue])

  return (
    <S.CardList>
      {showIllustration ? (
        <IllustrationCard
          notFound={contactsFiltered?.length === 0}
          isEmpty={
            contactsQuery?.data?.length === 0 ||
            contactsQuery?.data === undefined
          }
        />
      ) : (
        contactsFiltered?.map((contact) => (
          <ContactCard
            key={contact.phone + contact.id}
            id={contact.id}
            phone={contact.phone}
            name={contact.name}
            email={contact.email}
            navigateAction={handleNavigate}
          />
        ))
      )}

      <FloatingActionButton />
    </S.CardList>
  )
}
