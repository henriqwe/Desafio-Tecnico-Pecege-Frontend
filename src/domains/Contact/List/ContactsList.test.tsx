import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import { ContactList } from './ContactsList'
import { TestProviders } from '&utils/tests/TestProviders'

describe('ContactList', () => {
  it('should render the contact list correctly', async () => {
    const { container, findAllByTestId } = render(
      <TestProviders>
        <ContactList ascendentName={true} nameValue="" />
      </TestProviders>
    )
    const names = await findAllByTestId('name')
    const phones = await findAllByTestId('phone')
    const emails = await findAllByTestId('email')

    expect(names[0].textContent).toEqual('Chelsey Dietrich')
    expect(phones[0].textContent).toEqual('(254)954-1289')
    expect(emails[0].textContent).toEqual('Lucio_Hettinger@annie.ca')

    expect(container).toMatchSnapshot()
  })
})
