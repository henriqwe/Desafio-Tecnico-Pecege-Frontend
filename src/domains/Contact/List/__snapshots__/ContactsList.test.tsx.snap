// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactList should render the contact list correctly 1`] = `
<div>
  <div
    class="sc-jrkPcm efXjYQ"
  >
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              C
              D
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Chelsey Dietrich
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                (254)954-1289
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Lucio_Hettinger@annie.ca
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              C
              D
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Clementina DuBuque
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                024-648-3804
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Rey.Padberg@karina.biz
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              C
              B
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Clementine Bauch
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                1-463-123-4447
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Nathan@yesenia.net
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              E
              H
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Ervin Howell
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                010-692-6593 x09125
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Shanna@melissa.tv
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              G
              R
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Glenna Reichert
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                (775)976-6794 x41206
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Chaim_McDermott@dana.io
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              K
              W
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Kurtis Weissnat
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                210.067.6132
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Telly.Hoeger@billy.biz
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              L
              G
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Leanne Graham
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                1-770-736-8031 x56442
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Sincere@april.biz
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              M
              D
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Mrs. Dennis Schulist
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                1-477-935-8478 x6430
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Karley_Dach@jasper.info
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              N
              R
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Nicholas Runolfsdottir V
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                586.493.6943 x140
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Sherwood@rosamond.me
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0; transform: translateX(-100px) translateZ(0);"
    >
      <div
        class="sc-grXZZQ gZllkS"
        data-testid="contactcard-container"
      >
        <div
          class="sc-dkjKgF dvdRDZ"
        >
          <div
            class="sc-hLclGa kwlGWp"
          >
            <div
              class="sc-bcPKhP LlyrL"
              data-testid="avatar-container"
              variant="sm"
            >
              P
              L
            </div>
          </div>
          <div
            class="sc-eDnVMP bgmlqB"
          >
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-jSwlEQ hbBJZR"
                data-testid="name"
              >
                Patricia Lebsack
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="phone"
              >
                493-170-9623 x156
              </span>
            </div>
            <div
              class="sc-pqitP cjOBtN"
            >
              <span
                class="sc-gKHVLF dipxwJ"
                data-testid="email"
              >
                Julianne.OConner@kory.org
              </span>
            </div>
          </div>
        </div>
        <div
          class="sc-iBAaJG kwpJTv"
        >
          <svg
            aria-hidden="true"
            class="sc-ftLKQv jnvNod"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8.25 4.5l7.5 7.5-7.5 7.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      style="width: 100%; height: 100%; opacity: 0.2877594579822622;"
    >
      <div
        class="sc-bBbNsw feCeHL"
        data-testid="floatingactionbutton-container"
        position="right"
        title="Ir para mapa"
      >
        <div
          class="sc-ivDvhZ kSzkOs"
        >
          <div
            class="sc-cwKisF eXRRak"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                color="white"
                d="M12 18.5L9 17l-6 3V7l6-3l6 3l6-3v7M9 4v13m6-10v5m6.121 8.121a3 3 0 1 0-4.242 0c.418.419 1.125 1.045 2.121 1.879c1.051-.89 1.759-1.516 2.121-1.879zM19 18v.01"
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
